packages/react-reconciler/src/ReactFiberNewContext.js

```diff
@@ -90,17 +95,206 @@ function popProvider(providerFiber: Fiber): void {
+          break;
+        }
+        // No more siblings. Traverse up.
+        nextFiber = nextFiber.return;
+      }
+    }
+    fiber = nextFiber;
+  }
 }
 
-function getContextChangedBits(context: ReactContext<any>): number {
-  return isPrimaryRenderer ? context._changedBits : context._changedBits2;
+export function prepareToReadContext(
+  workInProgress: Fiber,
+  renderExpirationTime: ExpirationTime,
+): boolean {
+  currentlyRenderingFiber = workInProgress;
+  memoizedFirstContextDependency = workInProgress.firstContextDependency;
+  if (memoizedFirstContextDependency !== null) {
+    // Reset the work-in-progress list
+    workInProgress.firstContextDependency = null;
+
+    // Iterate through the context dependencies and see if there were any
+    // changes. If so, continue propagating the context change by scanning
+    // the child subtree.
+    let dependency = memoizedFirstContextDependency;
+    let hasPendingContext = false;
+    do {
+      const context = dependency.context;
+      const changedBits = isPrimaryRenderer
+        ? context._changedBits
+        : context._changedBits2;
+      if (changedBits !== 0) {
+        // Resume context change propagation. We need to call this even if
+        // this fiber bails out, in case deeply nested consumers observe more
+        // bits than this one.
+        propagateContextChange(
+          workInProgress,
+          context,
+          changedBits,
+          renderExpirationTime,
+        );
+        if ((changedBits & dependency.observedBits) !== 0) {
+          hasPendingContext = true;
+        }
+      }
+      dependency = dependency.next;
+    } while (dependency !== null);
+    return hasPendingContext;
+  } else {
+    return false;
+  }
 }
 
-export {
-  pushProvider,
-  popProvider,
-  getContextCurrentValue,
-  getContextChangedBits,
-};
+export function readContext<T>(
+  context: ReactContext<T>,
+  observedBits: void | number | boolean,
+): T {
+  invariant(
+    currentlyRenderingFiber !== null,
+    'Context.unstable_read(): Context can only be read while React is ' +
+      'rendering, e.g. inside the render method or getDerivedStateFromProps.',
+  );
+
+  if (typeof observedBits !== 'number') {
+    if (observedBits === false) {
+      // Do not observe updates
+      observedBits = 0;
+    } else {
+      // Observe all updates
+      observedBits = maxSigned31BitInt;
+    }
+  }
+
+  if (currentlyRenderingFiber.firstContextDependency === null) {
+    // This is the first dependency in the list
+    currentlyRenderingFiber.firstContextDependency = lastContextDependency = {
+      context: ((context: any): ReactContext<mixed>),
+      observedBits,
+      next: null,
+    };
+  } else {
+    invariant(
+      lastContextDependency !== null,
+      'Expected a non-empty list of context dependencies. This is likely a ' +
+        'bug in React. Please file an issue.',
+    );
+    if (lastContextDependency.context === context) {
```

Comment:
`lastContextDependency.context` should be stored separately in the execution environment so that it doesn't need to be accessed through a property. It can be null by default.

This check should be moved to the top so that we can bail as early as possible. You know that it `currentlyRenderingFiber.firstContextDependency === null` is not null if it matches.